### PR TITLE
Removed 'race: limit on 8192 simultaneously alive goroutines is excee…

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
…ded' error

address: https://github.com/istio/istio/issues/3789

### CODE REVIEW

File: mixer/adapter/redisquota/redisquota_integration_test.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ spec:
     maxAmount: 50
     validDuration: 30s
     bucketDuration: 1s
-    rateLimitAlgorithm: __RATE_LIMIT_ALGORITHM__
```
Comment (xiaolanz) [inline mixer/adapter/redisquota/redisquota_integration_test.go]: is it ok that the rolling window algorithm is dropped ? 
also other testcases?

File: mixer/adapter/redisquota/redisquota_integration_test.go [hunk 2]
```diff
@@ -107,149 +107,93 @@ spec:
...
-			},
-			statusCode: 0,
-		},
-		"Dimension override with best effort": {
-			attrs: map[string]interface{}{
-				"source.service":      "reviews",
-				"destination.service": "ratings",
-			},
-			quotas: map[string]istio_mixer_v1.CheckRequest_QuotaParams{
-				"overridden": {
```
Comment (xiaolanz) [inline mixer/adapter/redisquota/redisquota_integration_test.go]: overridden case wasn't tested.

File: mixer/adapter/redisquota/redisquota_integration_test.go [hunk 3]
```diff
@@ -107,149 +107,93 @@ spec:
...
+	}
 
-		serviceCfg := adapterConfig
-		serviceCfg = strings.Replace(serviceCfg, "__RATE_LIMIT_ALGORITHM__", algorithm, -1)
-		serviceCfg = strings.Replace(serviceCfg, "__REDIS_SERVER_ADDRESS__", mockRedis.Addr(), -1)
-		var cerr error
-		if args.ConfigStore, cerr = storetest.SetupStoreForTest(globalConfig, serviceCfg); cerr != nil {
-			t.Fatal(cerr)
-		}
+	mixerServer, err := server.New(args)
```
Comment (xiaolanz) [inline mixer/adapter/redisquota/redisquota_integration_test.go]: alternative may be start the mixer once, and run all testcases as requests?

File: mixer/adapter/redisquota/redisquota_integration_test.go (hunk 1 above)
Comment (mangchiandjjoe) [reply mixer/adapter/redisquota/redisquota_integration_test.go]: All functionalities were tested on the unit tests.
This integration test is designed to test the adapter is working with the mixer.
I think it is good.

File: mixer/adapter/redisquota/redisquota_integration_test.go (hunk 2 above)
Comment (mangchiandjjoe) [reply mixer/adapter/redisquota/redisquota_integration_test.go]: Same as above.

File: mixer/adapter/redisquota/redisquota_integration_test.go (hunk 3 above)
Comment (mangchiandjjoe) [reply mixer/adapter/redisquota/redisquota_integration_test.go]: It requires different configurations for rolling window and fixed window. 
I think for the integration test, this is good enough.

**Comment by ldemailly** [comment]
why are we using so many go routines ? maybe should address that instead ? can you file a followup

**Comment by mangchiandjjoe** [comment]
In the integration test, I created mixer instances with different configurations.
It was fine with go test without race option.
Three instances of mixer seem to be OK, but 4th instance killed the test with the race option.

**Comment by ldemailly** [comment]
my point is that's a lot of go routines, that it passes without race test doesn't mean it's ok to use so many go routines ... what are they all doing ?

**Comment by xiaolanz** [comment]
Can we git this merged?

**Comment by ldemailly** [comment]
anyone willing to take a stab at my question?

**Comment by xiaolanz** [comment]
@ldemailly this is an integration test that starts a entire mixer server. When we start a few servers at the same time, it exceeds goroutine budgets.  I wonder if you are asking to reduce goroutines in the mixer server. @ozevren

**Comment by ldemailly** [comment]
yes I am. maybe for production some hundred goroutine are fine but to process a few e2e tests request it should not consume that many. is there a breakdown?

**Comment by ozevren** [comment]
@ldemailly @xiaolanz Just had a brief offline chat with Joy. Mixer server will create goroutine pools internally (I believe the defaults are 1024 api worker + 1024 adapter worker, 2048 total each instantiation). If the server is not torn down correctly then these will linger. Joy mentioned that she tried adding shutdown in the test infrastructure loop, which sounds like the right thing to do. However, it looks like this caused some of the tests to fail? 

It looks like we should be investigating why that is happening.

**Comment by ldemailly** [comment]
ic, thanks ! where is the 1024(x2) number coming from ?
why not 64, why not 10000

and for testing can we set it to 10 or so?

**Comment by ozevren** [comment]
On 1024, I'll plead ignorance. Seems like a "good enough number".
Yes, the number is settable via Service Argument object.

**Comment by ldemailly** [comment]
filled #3948 we can continue there - thanks.

**Comment by mandarjog** [comment]
If we are starting several Mixers at the same time, we should start less number of go routines.

If we are running them one after another, then we should ensure that closing Mixer, actually releases all go routines.

File: root.crt [hunk 4]
```diff
@@ -0,0 +1,18 @@
+-----BEGIN CERTIFICATE-----
```
Comment (ldemailly) [inline root.crt]: where is this used/ how does it relate to this PR ?

**Comment by ldemailly** [comment]
ping

**Comment by costinm** [comment]
Please reopen if still needed, inactive for 1 month

**Pull request closed without merging**